UI & API for deleting a group

bgsavage opened the pull request:
![image](https://user-images.githubusercontent.com/5771116/29371805-34dbcd4e-825e-11e7-833b-5fc9ec438f50.png)

 - Requires GROUP_WRITE permission scoped to that school in order to delete.
 - Soft delete implemented as a restful PUT
 - New unit / integration tests still coming.

bgsavage committed:
Created modal for deleting a group.  In the sql, now only pulling back groups that have not been deleted.

bgsavage committed:
Merge branch 'develop' of https://github.com/SmarterApp/RDW_Admin into feature/deactivate-group

bgsavage committed:
Added API PUT method for groups which handles the soft delete of a group.  Uses the GROUP_WRITE permission to try to update the group.  If the group was not updated (either does not have permission or id does not exist), then will throw 404.  Handling error response in the groups modal.

bgsavage committed:
Now checking against GROUP_WRITE permission when returning the GroupSet to indicate to the UI whether or not this set of groups can be editable (or deleted) by the user.  Added exists method to the school repository to support this.

bgsavage committed:
Merge branch 'develop' of https://github.com/SmarterApp/RDW_Admin into feature/deactivate-group

bgsavage committed:
Updated failing unit / integration tests

bgsavage commented on webapp/src/main/java/org/opentestsystem/rdw/admin/model/GroupSet.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Set;
+
+public class GroupSet {
```
Had to create a wrapper which has the `writeable` property which indicates if the user is authorized to edit (and delete) this group.

bgsavage commented on webapp/src/main/java/org/opentestsystem/rdw/admin/model/GroupSet.java:
```diff
@@ -0,0 +1,46 @@
+    private boolean writeable;
+    private Set<Group> groups;
+
+    public boolean getWriteable() {
```
I need javadocs here.

bgsavage commented on webapp/src/main/java/org/opentestsystem/rdw/admin/repository/impl/JdbcGroupRepository.java:
```diff
@@ -57,8 +61,24 @@
+        );
+
+        if(result == 0)
+            throw new NoSuchElementException();
```
If nothing was updated, throwing an exception.  At this point, the UI shouldn't be showing the "Delete" link for groups the user cannot delete, so this shouldn't happen, but just in case.

bgsavage commented on webapp/src/main/java/org/opentestsystem/rdw/admin/repository/impl/JdbcGroupRepository.java:
```diff
@@ -79,6 +99,13 @@ private static Group mapResultToGroup(final ResultSet row) throws SQLException {
+    private static Map<String, Object> getGroupParameters(final Group group) {
+        return ImmutableMap.<String, Object>builder()
+                .put("id", group.getId())
+                .put("deleted", group.getIsDeleted())
```
I'm only implementing the the `deleted` field at this point.  When we come later (in phase 2?) to implement Edit, we can just include the other group fields once they've been validated.

bgsavage commented on webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultGroupService.java:
```diff
@@ -28,8 +30,15 @@
+        return GroupSet
+                .builder()
+                .groups(ImmutableSet.copyOf(repository.findAllByQuery(readPermission.getScope(), query))                )
+                .writeable(writePermission != null && schoolService.existsByPermissionScope(writePermission.getScope(), query.getSchoolId()))
```
It's possible for a user to have GROUP_READ for a group but not have GROUP_WRITE, so we need to make this check agains the given schoolId.

malaffoon commented on webapp/src/main/java/org/opentestsystem/rdw/admin/model/Group.java:
```diff
@@ -3,11 +3,20 @@
     private Subject subject;
     private int schoolYear;
     private int studentCount;
+    private boolean isDeleted;
```
convention in java would have this as `deleted` with the getter as `isDeleted`

bgsavage commented on webapp/src/main/resources/application.sql.yml:
```diff
@@ -14,21 +14,33 @@ sql:
         sc.name as school_name,
         sg.subject_id,
         sg.school_year,
+        sg.active,
```
Hm, I don't need active anymore.

malaffoon commented on webapp/src/main/java/org/opentestsystem/rdw/admin/model/Group.java:
```diff
@@ -44,16 +53,30 @@ public int getStudentCount() {
         private Subject subject;
         private int schoolYear;
         private int studentCount;
+        private boolean isDeleted;
```
not sure there is a convention for builders but i would make this `deleted` with the method `deleted` as well

bgsavage commented on webapp/src/main/resources/application.sql.yml:
```diff
@@ -14,21 +14,33 @@ sql:
+    update: >-
+      update student_group sg
+        join school s on sg.school_id = s.id
+      set sg.deleted = :deleted
```
Just re-iterating here, I'm only updating `deleted` field right now and should be updated once we implement edit groups.

malaffoon commented on webapp/src/main/java/org/opentestsystem/rdw/admin/repository/GroupRepository.java:
```diff
@@ -21,4 +21,11 @@
     List<Integer> findAllDistinctSchoolYears(@NotNull PermissionScope permissionScope);
+
+    /**
+     * Updates this group
```
suggest putting a comment about "just updating deleted flag" here

bgsavage commented on webapp/src/main/webapp/src/app/groups/delete-group.modal.ts:
```diff
@@ -0,0 +1,29 @@
+
+  save() {
+    this.group.isDeleted = true;
+    this.service.update(this.group).subscribe(res => {
```
UI implementation of "deleting a group".

esotrope replied:
I would think here that "update" is an implementation detail of soft delete and we would actually want to call this service method delete no?

bgsavage commented on webapp/src/main/webapp/src/app/groups/groups.component.html:
```diff
@@ -53,7 +53,13 @@
         </p-column>
         <p-column field="studentCount" header="{{'labels.groups.cols.student-count' | translate}}" [sortable]="true"></p-column>
-
+        <p-column *ngIf="isAuthorizedToWrite" field="isActive">
```
This field needs to be updated to isDeleted

malaffoon commented on webapp/src/main/java/org/opentestsystem/rdw/admin/service/SchoolService.java:
```diff
@@ -19,4 +20,12 @@
+     * @param schoolId          the school id to check
+     * @return true if the school exists and is part of the permission scope.
+     */
+    boolean existsByPermissionScope(@NotNull PermissionScope permissionScope, int schoolId);
```
Do we really need `ByPermissionScope` in the method name? Seems pretty obvious from the required parameter.

bgsavage commented on webapp/src/main/webapp/src/app/groups/groups.module.ts:
```diff
@@ -6,14 +6,22 @@ import { GroupService } from "./groups.service";
+    DeleteGroupModalComponent
+  ],
+  entryComponents: [
+    DeleteGroupModalComponent
```
This threw me off for a while when my modal wasn't working.  So for modal components we need to explicitly call it out in the `entryComponents` because it does not have a selector reference, is not part of a route definition, and is not part of bootstrap.  So the only way angular will know to add it to the componentFactory is if we call it out here in the `entryComponents`

bgsavage commented on webapp/src/main/webapp/src/app/groups/groups.service.ts:
```diff
@@ -51,15 +61,31 @@ export class GroupService {
+    apiModel.schoolName = uiModel.schoolName;
+    apiModel.subject = uiModel.subject;
+    apiModel.studentCount = uiModel.studentCount;
+    apiModel.isDeleted = uiModel.isDeleted;
```
I am sending the entire group object back to the API though, in anticipation that at some point the API will be saving all of this data.

bgsavage commented on webapp/src/main/webapp/src/app/shared/data/data.service.ts:
```diff
@@ -1,16 +1,13 @@
 @Injectable()
 export class DataService {
 
-  constructor(private http: Http, private staticDataService: CachingDataService) {
```
unrelated: not used

jtreuting commented on webapp/src/main/java/org/opentestsystem/rdw/admin/repository/impl/JdbcGroupRepository.java:
```diff
@@ -57,8 +61,24 @@
+    // TODO: This implementation only updates the isDeleted field.
+    // TODO: Other fields need to be validated before they can be updated.
+    @Override
+    public void update(@NotNull final PermissionScope permissionScope, @NotNull final Group group) {
```
Right now it seems the UI is setting the deleted flag and doing a normal update.  The soft-delete feels like an implementation detail where the front-end should just say it wants to delete it and the service deletes it.  It feels to me like these update methods should be delete and the API should use a DELETE verb, and then we just happen to be implementing that delete as a soft-delete by updating a specific flag in the DB.

bgsavage replied:
@malaffoon and I discussed this.  Since the UI will eventually have to become aware of the soft delete by allowing an "undelete", then a PUT would seem to work best since you can use that same PUT to do the undelete, instead of having to implement a `groups/id/undelete` as a separate api route.

bgsavage committed:
Renamed java from isDeleted -> deleted.
Added javadocs to GroupSet.
Removed unused sql select param.
Updated unit test to validate deleted and id fields.

bgsavage commented:
I'm closing this PR as it needs to be reworked since we should only ever be looking at the GROUP_WRITE permission and never the GROUP_READ permission in order to view groups.  This will simplify the API and UI we will no longer have to worry about the case of a user viewing a group that they don't have access to delete.

esotrope commented on webapp/src/main/java/org/opentestsystem/rdw/admin/model/Group.java:
```diff
@@ -3,11 +3,20 @@
 import org.opentestsystem.rdw.common.model.Subject;
 
 public class Group {
+    private int id;
```
If this is an entity ID, generally we have been using type long for good measure

bgsavage replied:
I can make this a long, though you know I really long for ints... :)

esotrope replied:
🌽 🧀 🌽 🙄

esotrope commented on webapp/src/main/resources/application.sql.yml:
```diff
@@ -40,6 +51,14 @@ sql:
+        select 1
+        from school
+        where id = :id
+          and (1=:statewide or district_id in (:district_ids) or id in (:school_ids))
```
```limit 1``` will short circuit this select I think

bgsavage replied:
Do we need to short circuit if I'm searching on a primary key?

esotrope replied:
Good point. You are just searching to match 1 indexed ID so it will make no difference 👍

esotrope commented on webapp/src/main/webapp/src/app/groups/delete-group.modal.html:
```diff
@@ -0,0 +1,23 @@
+      <button type="button" class="btn btn-default" (click)="cancel()">
+        {{ (unableToDelete? 'prompt.close' : 'labels.groups.delete.button-cancel') | translate }}
+      </button>
+      <button type="button" *ngIf="!unableToDelete" class="btn btn-primary" (click)="save()">
```
I think we might want to use btn-danger here but I'll delegate to the UI team to be sure

bgsavage replied:
I like how this looks, thanks.

@aaronsleeper thoughts??

![image](https://user-images.githubusercontent.com/5771116/29378872-8b187fce-8275-11e7-8733-79d516e96334.png)

aaronsleeper replied:
looks good
